File: weyl-core/src/main/java/io/quartic/weyl/core/model/Attributes.java
package io.quartic.weyl.core.model;

import com.fasterxml.jackson.annotation.JsonValue;
import io.quartic.common.SweetStyle;
import org.immutables.value.Value;

import java.util.Collections;
import java.util.Map;

@SweetStyle
@Value.Immutable
public interface Attributes {
    Attributes EMPTY_ATTRIBUTES = Collections::emptyMap;

    @JsonValue
    Map<AttributeName, Object> attributes();
}

File: weyl-core/src/main/java/io/quartic/weyl/core/model/Attributes.kt
package io.quartic.weyl.core.model

import com.fasterxml.jackson.annotation.JsonValue

interface Attributes {
    @get:JsonValue
    val attributes: Map<AttributeName, Any>

    companion object {
        val EMPTY_ATTRIBUTES = NaiveAttributes(emptyMap())

        data class NaiveAttributes(override val attributes: Map<AttributeName, Any>) : Attributes

        fun of(map: Map<AttributeName, Any>) = NaiveAttributes(map)
    }
}